lib/shared/s3-packages.js

```diff
@@ -0,0 +1,195 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const semver = require('semver');
+const Bluebird = require('bluebird');
+const request = Bluebird.promisifyAll(require('request'));
+const xml = Bluebird.promisifyAll(require('xml2js'));
+const information = require('./information');
+
+/**
+ * @summary Etcher S3 bucket URLS
+ * @namespace BUCKET_URL
+ * @public
+ */
+exports.BUCKET_URL = {
+
+  /**
+   * @property {String} PRODUCTION
+   * @memberof BUCKET_URL
+   * @description
+   * Etcher production S3 bucket URL
+   */
+  PRODUCTION: 'https://resin-production-downloads.s3.amazonaws.com',
+
+  /**
+   * @property {String} SNAPSHOT
+   * @memberof BUCKET_URL
+   * @description
+   * Etcher snapshot S3 bucket URL
+   */
+  SNAPSHOT: 'https://resin-nightly-downloads.s3.amazonaws.com'
+
+};
+
+/**
+ * @summary Etcher S3 package name
+ * @constant
+ * @private
+ * @type {String}
+ */
+const S3_PACKAGE_NAME = 'etcher';
+
+/**
+ * @summary Number of packages per Etcher version
+ * @constant
+ * @private
+ * @type {Number}
+ */
+const NUMBER_OF_PACKAGES = 8;
+
+/**
+ * @summary Get the correct S3 bucket url from a release type
+ * @function
+ * @public
+ *
+ * @param {RELEASE_TYPE} releaseType - release type
+ * @returns {(String|Null)} S3 bucket url
+ *
+ * @example
+ * const bucketUrl = s3Packages.getBucketUrlFromReleaseType(information.RELEASE_TYPE.PRODUCTION);
+ *
+ * if (bucketUrl) {
+ *   console.log(bucketUrl);
+ * }
+ */
+exports.getBucketUrlFromReleaseType = (releaseType) => {
+  if (releaseType === information.RELEASE_TYPE.PRODUCTION) {
+    return exports.BUCKET_URL.PRODUCTION;
+  }
+
+  if (releaseType === information.RELEASE_TYPE.SNAPSHOT) {
+    return exports.BUCKET_URL.SNAPSHOT;
+  }
+
+  return null;
+};
+
+/**
+ * @summary Get all remote versions from an S3 bucket
+ * @function
+ * @private
+ *
+ * @description
+ * We memoize based on the assumption that the received latest version
+ * number will not increase while the application is running.
+ *
+ * @param {String} bucketUrl - s3 bucket url
+ * @fulfil {String[]} - remote versions
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getRemoteVersions(s3Packages.BUCKET_URL.PRODUCTION).then((versions) => {
+ *   _.each(versions, (version) => {
+ *     console.log(version);
+ *   });
+ * });
+ */
+exports.getRemoteVersions = _.memoize((bucketUrl) => {
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return request.getAsync(bucketUrl)
```

Comment:
`getAsync` will fail if so, but I'll catch it myself so we can provide a more meaningful error messsage.